meta-balena-common/recipes-containers/hostapp-update/files/hostapp-update

```diff
@@ -37,19 +45,235 @@ run_current_hooks_and_recover () {
+		if [ -z "${supervisor_version}" ]; then
+			ERROR "Could not get current supervisor version from the API, got ${resp}"
+			return 1
+		fi
+		scheduled_supervisor_version=$(echo "${resp}" | jq -e -r '.d[0].should_be_managed_by__supervisor_release[0].supervisor_version' | tr -d 'v')
+		if [ -n "${scheduled_supervisor_version}" ] && [ "${scheduled_supervisor_version}" != "null" ]; then
+			if version_gt "${scheduled_supervisor_version}" "${supervisor_version}"; then
+				# The supervisor is scheduled to update
+				echo "${supervisor_version} ${scheduled_supervisor_version}"
+				return 0
+			fi
+		fi
+		echo "${supervisor_version} ${supervisor_version}"
+	else
+		ERROR "Could not fetch current supervisor version from the API, got ${resp}"
+		return 1
+	fi
+}
+
+#######################################
+# Helper function to patch the supervisor version in the target state.
+# Globals:
+#  API_ENDPOINT
+#  APIKEY
+#  UUID
+#  SLUG
+#
+# Arguments:
+#  version: supervisor version to update the target state to
+# Returns
+#  0: Success
+#  1: Failure
+#######################################
+_patch_supervisor_version() {
+	local version=$1
+	local current_version
+	local _status_code
+	local _errfile
+	local _outfile
+	local UPDATER_SUPERVISOR_TAG
+	local UPDATER_SUPERVISOR_ID
+
+	[ -z "${version}" ] && INFO "Supervisor version is required" && return 1
+	UPDATER_SUPERVISOR_TAG="v${version}"
+
+	# Get the supervisor id
+	resp=$(${CURL} --header "Authorization: Bearer ${APIKEY}" "${API_ENDPOINT}/v5/supervisor_release?\$select=id,image_name&\$filter=((device_type%20eq%20'$SLUG')%20and%20(supervisor_version%20eq%20'${UPDATER_SUPERVISOR_TAG}'))")
+	if UPDATER_SUPERVISOR_ID=$(echo "${resp}" | jq -e -r '.d[0].id'); then
+		INFO "Extracted supervisor vars: ID: $UPDATER_SUPERVISOR_ID"
+		INFO "Setting supervisor version in the API..."
+
+		_errfile=$(mktemp)
+		_outfile=$(mktemp)
+		if _status_code=$(${CURL} --request PATCH -w "%{http_code}" --show-error -o "${_outfile}" --header "Authorization: Bearer ${APIKEY}" --header 'Content-Type: application/json' "${API_ENDPOINT}/v6/device(uuid='${UUID}')" --data-binary "{\"should_be_managed_by__supervisor_release\": \"${UPDATER_SUPERVISOR_ID}\"}" 2> "${_errfile}"); then
+			rm -f "${_errfile}"
+			case "${_status_code}" in
+				2*) INFO "Successfully set supervision version in target state";rm -f "${_outfile}";return 0;;
+				4*) WARN "[${_status_code}]: Bad request: $(cat "${_outfile}")"; rm -f "${_outfile}"; if current_version=$(_fetch_supervisor_version | cut -d " " -f1); then if version_gt "${current_version}" "${version}"; then return 0; else return 1; fi; else return 1; fi;;
+				*) WARN "[${_status_code}]: Request failed: $(cat "${_outfile}")";rm -f "${_outfile}";return 1;;
+			esac
+		else
+			WARN "$(cat "${_errfile}")"
+			rm -f "${_errfile}"
+			return 1
+		fi
+		else
+			WARN "Failed fetching supervisor id from API: ${resp}"
+			return 1
+		fi
+}
```

Comment:
I'd probably like to understand what is the rational for the hostOS having to call the API for this and whether whatever is triggering the hostOS update should do this instead.

Reply:
Replied at https://balena.zulipchat.com/#narrow/channel/346007-balena-io.2FbalenaCloud/topic/Roadmap.20to.20tri-app/near/486479671